src/DistBCSR.hpp

```diff
@@ -0,0 +1,1192 @@
+    c_dbcsr_read_d(&matrix,cfname,fdist);
+  }
+
+  inline void write(dm_dbcsr& matrix, char* cfname){
+    c_dbcsr_write_d(&matrix,cfname);
+  }
+
+  inline void maxabs(dm_dbcsr& matrix, double& amv){
+    c_dbcsr_maxabs_d(&matrix,&amv);
+  }
+
+}
+
+
+/*
+ * Class to hold MPI-setup for DBCSR
+ */
+class DBCSR_Environment {
+  public:
+    std::string env_id;
+    int dbcsr_dims[2];
+    int dbcsr_periods[2];
+    int dbcsr_coords[2];
+    int dbcsr_reorder;
+    void* dbcsr_dist;
+    MPI_Comm dbcsr_group;
+    std::vector<int> dbcsr_row_dist;
+    std::vector<int> dbcsr_col_dist;
+
+    int mpi_rank;
+    int mpi_size;
+    int nblk_row;
+    int nblk_col;
+
+    DBCSR_Environment(){
+      env_id = "invalid";
+    };
+
+    DBCSR_Environment(int nblk_row_in, int nblk_col_in, const std::string& id_in="dbcsr_env_default"){
+      this->init(nblk_row_in,nblk_col_in,id_in);
+    };
+
+    void init(int nblk_row_in, int nblk_col_in, const std::string& id_in="dbcsr_env_default"){
+      env_id = id_in;
+      nblk_row = nblk_row_in;
+      nblk_col = nblk_col_in;
+      MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+      MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+
+      // Make 2D grid
+      dbcsr_dims[0] = 0;
+      dbcsr_dims[1] = 0;
+      MPI_Dims_create(mpi_size, 2, dbcsr_dims);
+      dbcsr_periods[0] = 1;
+      dbcsr_periods[1] = 1;
+      dbcsr_reorder = 0;
+      MPI_Cart_create(MPI_COMM_WORLD, 2, dbcsr_dims, dbcsr_periods, dbcsr_reorder, &dbcsr_group);
+
+      MPI_Cart_coords(dbcsr_group, mpi_rank, 2, dbcsr_coords);
+
+      c_dbcsr_init_lib();
+
+      dbcsr_row_dist.resize(nblk_row);
+      dbcsr_col_dist.resize(nblk_col);
+
+      for(int i=0; i < nblk_row; i++){
+       dbcsr_row_dist[i] = (i+1) % dbcsr_dims[0];
+      }
+      for(int i=0; i < nblk_col; i++){
+       dbcsr_col_dist[i] = (i+1) % dbcsr_dims[1];
+      }
+
+      dbcsr_dist = nullptr;
+
+      int dr = (int)dbcsr_row_dist.size();
+      int dc = (int)dbcsr_col_dist.size();
+      MPI_Fint fcomm = MPI_Comm_c2f(dbcsr_group);
+      c_dbcsr_distribution_new_aux(&dbcsr_dist, &fcomm,
+                                   &dbcsr_row_dist[0], dr,
+                                   &dbcsr_col_dist[0], dc);
+
+    };
+
+    ~DBCSR_Environment(){
+      this->free();
+    };
+
+    void free(){
+      if (dbcsr_dist == nullptr) return;
+      c_dbcsr_distribution_release(&dbcsr_dist);
+      dbcsr_dist = nullptr;
+      MPI_Fint fcomm = MPI_Comm_c2f(dbcsr_group);
+    //if (verbose)
+    //  c_dbcsr_finalize_lib_aux(&fcomm);
+    //else
+        c_dbcsr_finalize_lib_aux_silent(&fcomm);
+      MPI_Comm_free(&dbcsr_group);
+    };
+
+};
+
+/*
+ * Wrapper-class for dbcsr
+ */
+class DistBCSR {
+
+  private:
+    DBCSR_Environment* dbcsr_env;
+
+    dm_dbcsr dbcsr_matrix;
+    std::string mname;
+    size_t nrow;
+    size_t ncol;
+    std::vector<int> row_dims;
+    std::vector<int> col_dims;
+    double dbcsr_thresh;
+
+  public:
+    DistBCSR(){
+      dbcsr_matrix = nullptr;
```

Comment:
it would help reading and documenting the code if you separate declaration from definition within the same header.

Reply:
It certainly will, but as a first step I wanted to keep my addition compact. It'll also be the first C++-source in the library (I assume), which may also require an overhaul of some make/cmake files.
@dev-zero  I suppose the maintainers will be better suited to make these modifications.

Reply:
We already have CUDA files and also C++ already in the tree (as part of the CUDA SMM library), C++ sources are therefore no problem at all.

Reply:
> It certainly will, but as a first step I wanted to keep my addition compact

it would be as compact as now, I only suggested to separate definitions from declarations and still keep things in one header.

Reply:
> I'm not even sure how to do this while avoiding multiple definitions.

You are right.